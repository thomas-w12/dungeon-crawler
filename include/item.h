#ifndef ITEM_H
#define ITEM_H

#define ITEM_TYPES (const char*[]){"sword", "stone", "shield", "ladder"}
#define ITEM_DESCRIPTIONS (const char*[]){"Fight off monsters", "Stone the giant", "Shield against trap", "Climb out of trap"}

typedef struct Item{
    // Item type and description depends on the item id
    int ID;
    char type[10];
    char description[20];
} Item;

Item* Item_construct(int ID);
void displayItem(Item* item);
void displayItems(Item* items[], int itemCount);
void freeItem(Item* item);
void freeItems(Item* items[], int* itemsCount);
// void addItemToRoom(Room* room, Item item);
// void addItemToPlayer(Player* player, Item item);
// void removeItemFromRoom(Room* room, Item item); // Remove item after player collects item
// void removeItemFromPlayer(Player* player, Item item); // Remove item after player uses them or dies

#endif