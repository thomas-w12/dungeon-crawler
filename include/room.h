#ifndef ROOM_H
#define ROOM_H

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "item.h"
#include "command_parser.h"
#include "event.h"
#include "queue.h"

#define MAX_ROOMS 10
#define MAX_ROOM_LINE_LEN 100 // Max length of the room line stored in the layout.txt
#define MAX_ITEMS_IN_ROOM 3
#define MAX_ROOM_EVENTS 3
#define MAX_DIRECTIONS 4
#define PROB_ITEM_IN_ROOM 0.2 // Probability of having an item in a room
#define PROB_EVENT_IN_ROOM 0.1 // Probability of having an event in a room
#define EXPAND_ROOM_COUNT 20

typedef enum DirectionsIndex{
    NorthIndex,
    SouthIndex,
    WestIndex,
    EastIndex
} DirectionsIndex;

typedef struct Room{
    int ID; 
    struct Room *north;
    struct Room *south;
    struct Room *east;
    struct Room *west;
    EventNode* events; //points to the first event which holds the rest of list
    ItemNode* items;
} Room;

void Room_destroy(Room* room);
Room* Room_construct(int ID, Room* north, Room* south, Room* west, Room* east, EventNode* events, ItemNode* items);
void reallocRooms(Room*** roomsPtr, int* allocRoomsSize, int expectedSize);
void expandRoom(Room*** roomsPtr, Room* firstRoom, int* roomCount, int noRoomsToAdd, int* allocRoomsSize);
void generateLayout(Room*** roomsPtr, int* roomCount, int noRoomsToAdd, int* allocRoomsSize);
void displayRoom(Room* room);
void displayRooms(Room** rooms, int roomCount);
void freeRooms(Room*** roomsPtr, int* roomCount);

#endif