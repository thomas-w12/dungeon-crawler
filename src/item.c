#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "../include/item.h"
#include "../include/global.h"


Item* Item_construct(int ID) {
    Item* newItem = malloc(sizeof(Item));
    if (!newItem) {
        printf("Error: Could not allocate memory for Item.\n");
        return NULL;
    }
    newItem->ID = ID;
    strcpy(newItem->type, ITEM_TYPES[ID]);
    strcpy(newItem->description, ITEM_DESCRIPTIONS[ID]);
    return newItem;
}

ItemNode* ItemNode_construct(Item* item){
    ItemNode* itemNode = malloc(sizeof(ItemNode));
    if (itemNode == NULL){
        perror("Could not allocate memory for ItemNode");
        return NULL;
    }

    itemNode->data = item;
    itemNode->next = NULL;

    return itemNode;
}

ItemNode* ItemList_insert(ItemNode** head, Item* item){
    ItemNode* itemNode = ItemNode_construct(item);
    if (itemNode == NULL){
        perror("Failed Item List insert");
        return NULL;
    }
    // cannot have duplicate events
    if (*head == NULL){
        *head = itemNode;
    }else{
        ItemNode* curr = *head;

        while(curr->next != NULL){
            if (curr->data->ID == item->ID){
                perror("Duplicate items are not allowed");
                return NULL;
            }
            curr = curr->next;
        }

        curr->next = itemNode;
    }
    return itemNode;
}

Item* ItemList_deleteAtIndex(ItemNode** head, int itemIndex){
    Item* item = NULL;

    if (*head == NULL){
        return item;
    }else if (itemIndex == 0){// remove head
        if ((*head)->next == NULL){
            ItemNode* temp = (*head);
            item = (*head)->data;
            *head = NULL;
            free(temp);
            return item;
        }
        ItemNode* temp = (*head)->next;
        item = (*head)->data;
        free(*head);
        *head = temp;
    }else{
        ItemNode* prev = *head;
        ItemNode* curr = (*head)->next;

        int count = 1; //because we are not staring from the beginnig
        while((curr != NULL) && (itemIndex > count)){
            prev = curr;
            curr = curr->next;
            count++;
        }
        
        if (curr != NULL){
            prev->next = curr->next;
            item = curr->data;
            free(curr);
        }
        displayItem(item);
    }
    return item;
}

Item* ItemList_deleteItemID(ItemNode** head, int itemID){
    Item* item = NULL;

    if (*head == NULL){
        return item;
    }else if (itemID == (*head)->data->ID){// remove head
        if ((*head)->next == NULL){
            ItemNode* temp = (*head);
            item = (*head)->data;
            *head = NULL;
            free(temp);
            return item;
        }
        ItemNode* temp = (*head)->next;
        item = (*head)->data;
        free(*head);
        *head = temp;
    }else{
        ItemNode* prev = *head;
        ItemNode* curr = (*head)->next;

        int count = 1; //because we are not staring from the beginnig
        while((curr != NULL) && (itemID != curr->data->ID)){
            prev = curr;
            curr = curr->next;
            count++;
        }
        
        if (curr != NULL){
            prev->next = curr->next;
            item = curr->data;
            free(curr);
        }
    }
    return item;
}

int itemListCount(ItemNode* head){
    ItemNode* curr = head;
    int count = 0;
    while (curr != NULL){
        curr = curr->next;
        count++;
    }
    return count;
}

void itemListToArr(ItemNode* head, Item* arr[], int arrSize){
    ItemNode* curr = head;
    for (int i=0; i<arrSize; i++){
        if (curr == NULL){
            arr[i] = NULL;
        }else{
            arr[i] = curr->data;
            curr = curr->next;
        }
    }
}

void displayItem(Item* item){
    if (item != NULL) {
        printf("\nItem - ID: %d\tType: %s\tDescription: %s", item->ID, item->type, item->description);
    }else{
        printf("Item is NULL");
    }
}

void displayItems(Item* items[], int itemCount, int maxItems){
    if (itemCount == 0){
        printf("There are no items!");
        return;
    }

    for (int i=0; i<maxItems; i++){
        displayItem(items[i]);
    }
}

void printItemList(ItemNode* head){
    if (head == NULL){
        printf("There are no items!");
        return;
    }
    ItemNode* curr = head;
    while (curr != NULL){
        // printf("%d->", curr->data);
        displayItem(curr->data);
        curr = curr->next;
    }
}

void freeItem(Item* item){
    if (item == NULL) {
        return;
    }
    free(item);
}

void freeItemNode(ItemNode* itemNode){
    if (itemNode == NULL) return;
    freeItemNode(itemNode->next);
    free(itemNode->data);
    free(itemNode);
}

void freeItemList(ItemNode** headPtr){
    if (*headPtr == NULL) return;
    freeItemNode(*headPtr);
    *headPtr = NULL;
}

void freeItems(Item* items[], int* itemsCount){
    for (int i=0; i<*itemsCount; i++){
        free(items[i]);
    }
    *itemsCount = 0;
}