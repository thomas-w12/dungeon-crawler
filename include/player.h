#ifndef PLAYER_H
#define PLAYER_H

#include <stdio.h>
#include "item.h"
#include "room.h"
// struct Room;

#define MAX_PLAYER_ITEMS 3

typedef struct Player {
    char name[10]; // Tell player to enter name at begining. Max length of about 10 characters
    int currentRoom; //ID of current room
    int health; // Health bar
    int score;
    Item* inventory[MAX_PLAYER_ITEMS]; // List of items the player has
    int itemsCount;
} Player;

void moveNorth(Player* player, Room* currentRoom);
void moveSouth(Player* player, Room* currentRoom);
void moveWest(Player* player, Room* currentRoom);
void moveEast(Player* player, Room* currentRoom);
void pickUpItem(Player* player, Item* item);

#endif