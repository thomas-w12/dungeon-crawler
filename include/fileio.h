#ifndef FILEIO_H
#define FILEIO_H

#include <stdio.h>
#include "room.h"
#include "player.h"

#define DNE -1

int saveLayout(const char* layoutStateFPath, Room** rooms, int roomCount);
int loadLayout(const char* layoutStateFPath, Room** rooms, int* roomCount);
int savePlayerState(const char* playerStateFPath, Player* player);
int loadPlayerState(const char* playerStateFPath, Player* player);

#endif
