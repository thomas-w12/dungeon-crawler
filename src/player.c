#include <ctype.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "../include/room.h"
#include "../include/item.h"
#include "../include/player.h"
#include "../include/npc.h"
#include "../include/global.h"


Player* Player_construct(char* name, int currentRoom, int health, int score, Room* currentRoomPtr){
    Player* player = (Player*)malloc(sizeof(Player));
    if (player == NULL){
        printf("Could not create player\n");
        return NULL;
    }

    player->name = strdup(name);
    if (player->name == NULL){
        printf("Could not create player name\n");
        free(player);
        return NULL;
    }
    player->currentRoom = currentRoom;
    player->health = health;
    player->score = score;
    
    player->inventory = NULL;
    player->currentRoomPtr = currentRoomPtr;

    return player;
}

void displayPlayer(Player* player){
    printf("Player - Name: %s\tCurrent Room: %d\tHealth: %d\tScore: %d\n", player->name, player->currentRoom, player->health, player->score);
    displayPlayerInventory(player);
}

void displayPlayerInventory(Player* player){
    printf("Items in inventory (%d): \n", itemListCount(player->inventory));
    printItemList(player->inventory);
}

void updatePlayerRoom(Player* player, Room* currentRoom){
    player->currentRoom = currentRoom->ID;
    player->currentRoomPtr = currentRoom;
}

void pickUpItem(Player* player, int pickupItemID){
    if (itemListCount(player->inventory) >= MAX_PLAYER_ITEMS){
        printf("Inventory is full. Cannot pick up item.\n");
        return;
    }

    // Find the item with matching ID in the current room
    Item* pickupTargetItem = ItemList_deleteItemID(&player->currentRoomPtr->items, pickupItemID);
    if (pickupTargetItem == NULL) {
        printf("Invalid item ID\n");
        return;
    }

    // Add item to player inventory
    ItemList_insert(&player->inventory, pickupTargetItem);
    printf("You picked up an Item. Press i to see your inventory\n");
}

void dropItem(Player* player, int dropItemID){
    if (itemListCount(player->inventory) == 0){ // Redundant, should remove
        printf("Inventory is empty. Cannot drop item.\n");
        return;
    }

    // Find the item with matching ID in the player's inventory
    Item* dropTargetItem = ItemList_deleteItemID(&player->inventory, dropItemID);

    if (dropTargetItem == NULL) {
        printf("Invalid item ID\n");
        return;
    }

    // Add item to room
    ItemList_insert(&player->currentRoomPtr->items, dropTargetItem);
    printf("You dropped an Item. Press i to see your inventory\n");
}

bool useItem(Player* player, int itemID){
    if (itemListCount(player->inventory) == 0){ // Redundant, should remove
        printf("Inventory is empty. There is no item to use.\n");
        return false;
    }
    
    bool usedItem = false;

    switch((ItemsID) itemID){
        case POTION:
            if (ItemList_deleteItemID(&player->inventory, itemID) == NULL){
                printf("You don't have any %s.\n", ITEM_TYPES[(int) POTION]);
            }else{
                increasePlayerHealth(player, 50);
                usedItem = true;
            }
            break;
        default:
            printf("You can not use this item\n");
            break;
    }
    return usedItem;
}

void decreasePlayerHealth(Player* player, int damage){
    player->health -= damage;
    printf("Ouch you took %d%% damage!\n", damage);
}

void increasePlayerHealth(Player* player, int health){
    player->health = (player->health + health) > 100 ? 100 : player->health + health;
    printf("Your health increased by %d%%\n", health);
}

void increasePlayerScore(Player* player, int score){
    player->score += score;
}

void freePlayer(Player* player){
    if (player == NULL){
        return;
    }
    free(player->name);
    freeItemList(&player->inventory);
    free(player);
}

bool validateItemUse(Player* player, char message[], ItemsID validItemId){
    printf("%s", message);

    int item = parse_item_command();

    if (item == -1){
        printf("Understandable have a nice day!\n");
        return false;
    }else{
        if (item != (int) validItemId){
            printf("Wrong item!\nYou need a %s - %d\n", ITEM_TYPES[(int) validItemId], (int) validItemId);
            return validateItemUse(player, message, validItemId);
        }else if (ItemList_deleteItemID(&player->inventory, item) == NULL){
            printf("You dont have that item\nPress -1 to quit\n");
            printf("Your inventory(%d): \n", itemListCount(player->inventory));
            printItemList(player->inventory);
            return validateItemUse(player, message, validItemId);
        }else{
            return true;
        }
    }
}

bool triggerBlockedEvent(Room* room, Player* player){
    char choice = getUserInputCommand();
    switch(choice){
        case USE:
            if (validateItemUse(player, "Enter the id of the item you would like to use: ", PICKAXE) == false) return false;

            printf("You just broke throught the room, you can now enter it.\n");
            return true;
        case DECLINE:
            printf("Alright, room ignored\n");
            return false;
        default:
            printf("Invalid choice\nPress:\nU - use an item\nX - Ignore room\n");
            return triggerBlockedEvent(room, player);
    }
}

bool triggerLockedEvent(Room* room, Player* player){
    char choice = getUserInputCommand();
    switch(choice){
        case USE:
            if (validateItemUse(player, "Enter the id of the item you would like to use: ", KEY) == false) return false;

            printf("You just unlocked the room, you can now enter it.\n");
            return true;
        case DECLINE:
            printf("Alright, room ignored\n");
            return false;
        default:
            printf("Invalid choice\nPress:\nU - use an item\nX - Ignore room\n");
            return triggerLockedEvent(room, player);
    }
}

bool triggerPitTrapEvent(Player* player){
    if (ItemList_deleteItemID(&player->inventory, LADDER) != NULL){
        printf("\nYou just fell into a pit, but you climbed out of it with your ladder!\n");
    }else{
        printf("\nYou just fell into a pit.\n");
        decreasePlayerHealth(player, 100);
    }
    return true;
}

bool triggerPuzzleEvent(Room* room, Player* player, int puzzleIndex){
    char userAnswerBuff[100];
    scanf(" %99[^\n]", userAnswerBuff);
    clearBuffer();
    
    if (toupper((int)userAnswerBuff[0]) == (int) DECLINE){
        return false;
    }

    if (strcmp(userAnswerBuff, PUZZLE_ANSWERS[puzzleIndex]) == 0){
        printf("Correct answer, you can proceed to the room\n");
        return true;
    }else{
        printf("Wrong answer! \nTry again! or Enter 'X' to quit\n" );
        return triggerPuzzleEvent(room, player, puzzleIndex);
    }
}

bool triggerEvents(Room* room, Player* player){
    bool eventsPassed = true;
    
    EventNode* currEventNode = room->events;

    while (currEventNode != NULL){
        Event currEvent = currEventNode->data;
        eventsPassed &= triggerEvent(room, player, currEvent);

        if (eventsPassed == false) return eventsPassed; // Dont continue if one event fails
        if (player->health <= 0) return false;

        currEventNode = currEventNode->next;
    }

    printf("Events Passed: %d\n", eventsPassed);
    return eventsPassed;
}

bool triggerEvent(Room* room, Player* player, Event currEvent){
    bool eventPassed = false;
    switch(currEvent){
        case NORMAL:
            printf("This room is a normal room\n");
            EventList_delete(&room->events, currEvent);
            eventPassed = true;
            break;
        case BLOCKED:
            printf("This room is blocked, you need a pick axe to break through\nPress:\nU - use an item\nX - Ignore room\n");
            eventPassed = triggerBlockedEvent(room, player);
            if (eventPassed) EventList_delete(&room->events, currEvent);
            break;
        case LOCKED:
            printf("This room is locked, you need a key to open it\nPress:\nU - use an item\nX - Ignore room\n");
            eventPassed = triggerLockedEvent(room, player);
            if (eventPassed) EventList_delete(&room->events, currEvent);
            break;
        case FIRE_TRAP:
            printf("You just got caught in a fire trap!\n");
            decreasePlayerHealth(player, 25);
            eventPassed = true;
            break;
        case PIT_TRAP:
            eventPassed = triggerPitTrapEvent(player);
            break;
        case PUZZLE:
            printf("You have to solve this puzzle to continue to this room\n");
            int puzzleIndex = generateRandomInt(0, PUZZLE_COUNT-1);
            printf("%s\n", PUZZLE_QUESTIONS[puzzleIndex]);
            eventPassed = triggerPuzzleEvent(room, player, puzzleIndex);
            if (eventPassed) EventList_delete(&room->events, currEvent);
            break;
        case NPC_BOSS:
            break;
        case NPC_TRADE:
        case NPC_TALK:
            NPC npc = createNPC(currEvent);
            npc.interact(&npc, player);
            eventPassed = true;

            if (currEvent == NPC_BOSS) EventList_delete(&room->events, currEvent);
            break;
        case POISON_TRAP:
            printf("You just entered a poison room!\n");
            decreasePlayerHealth(player, 15);
            eventPassed = true;
            break;
    }

    return eventPassed;
}