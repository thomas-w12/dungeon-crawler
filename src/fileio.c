#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../include/fileio.h"
#include "../include/event.h"

int saveLayout(const char* layoutStateFPath, Room** rooms, int roomCount) {
    FILE* file = fopen(layoutStateFPath, "w");

    if (file == NULL) {
        perror("Error opening layout file");
        return EXIT_FAILURE;
    }

    fprintf(file, "%d\n", roomCount);

    for (int i = 0; i < roomCount; i++) {
        Room* room = rooms[i];
        fprintf(file, "%d,%s,%s,", room->ID, room->name, room->description);

        fprintf(file, "%d,%d,%d,%d,", 
                (room->north ? room->north->ID : DNE),
                (room->south ? room->south->ID : DNE),
                (room->west ? room->west->ID : DNE),
                (room->east ? room->east->ID : DNE));

        int arr[MAX_ROOM_EVENTS];
        eventListToArray(room->events, arr, MAX_ROOM_EVENTS);
        for (int j = 0; j < MAX_ROOM_EVENTS; j++) {
            fprintf(file, "%d,", arr[j]);
        }
        
        for (int j = 0; j < room->itemsCount; j++) {
            fprintf(file, "%d,", room->items[j]->ID);
        }
        fprintf(file, "\n");
    }

    fclose(file);
    return EXIT_SUCCESS;
}

int loadLayout(const char* layoutStateFPath, Room** rooms, int* roomCount) {
    FILE* file = fopen(layoutStateFPath, "r");

    if (file == NULL) {
        perror("Error opening layout file");
        return EXIT_FAILURE;
    }

    fscanf(file, "%d", roomCount);

    for (int i = 0; i < *roomCount; i++) {
        rooms[i] = (Room*)malloc(sizeof(Room));
        if (rooms[i] == NULL) {
            printf("Memory allocation failed for room %d\n", i);
            continue;
        }

        int id, northID, southID, westID, eastID;
        char name[10], description[30];

        fscanf(file, "%d,%10[^,],%30[^,],%d,%d,%d,%d,", 
               &id, name, description, &northID, &southID, &westID, &eastID); //30 here reads at most 30 characters as the description and 10 as the name

        rooms[i]->ID = id;
        strcpy(rooms[i]->name, name);
        strcpy(rooms[i]->description, description);
        rooms[i]->north = (northID != DNE) ? rooms[northID] : NULL;
        rooms[i]->south = (southID != DNE) ? rooms[southID] : NULL;
        rooms[i]->west = (westID != DNE) ? rooms[westID] : NULL;
        rooms[i]->east = (eastID != DNE) ? rooms[eastID] : NULL;
        rooms[i]->itemsCount = 0;

        rooms[i]->events = NULL;
        for (int j=0; j<MAX_ROOM_EVENTS; j++){
            int event;
            fscanf(file, "%d,", &event);
            if (event != DNE){
                rooms[i]->events = EventList_insert(rooms[i]->events, event);
            }
        }

        while (fgetc(file) != '\n' && !feof(file)) {
            fseek(file, -1, SEEK_CUR); // Go back one char from current position(fgetc pushed it forward)
            int itemID;
            fscanf(file, "%d,", &itemID);
            rooms[i]->items[rooms[i]->itemsCount] = Item_construct(itemID);
            rooms[i]->itemsCount++;
        }
    }

    fclose(file);
    return EXIT_SUCCESS;
}

int savePlayerState(const char* playerStateFPath, Player* player) {
    FILE* file = fopen(playerStateFPath, "w");

    if (file == NULL) {
        perror("Error opening player state file");
        return EXIT_FAILURE;
    }

    fprintf(file, "%s,%d,%d,%d,", player->name, player->currentRoom, player->health, player->score);
    for (int i = 0; i < MAX_PLAYER_ITEMS; i++) {
        if (player->inventory[i] != NULL) {
            fprintf(file, "%d,", player->inventory[i]->ID);
        }
    }

    fprintf(file, "\n");
    fclose(file);
    return EXIT_SUCCESS;
}

int loadPlayerState(const char* playerStateFPath, Player* player) {
    FILE* file = fopen(playerStateFPath, "r");

    if (file == NULL) {
        perror("Error opening player state file");
        return EXIT_FAILURE;
    }

    fscanf(file, "%9[^,],%d,%d,%d,", player->name, &player->currentRoom, &player->health, &player->score);
    player->itemsCount = 0;

    while (fgetc(file) != '\n' && !feof(file)) {
        int itemID;
        fscanf(file, "%d,", &itemID);
        player->inventory[player->itemsCount++] = Item_construct(itemID);
    }

    fclose(file);
    return EXIT_SUCCESS;
}
