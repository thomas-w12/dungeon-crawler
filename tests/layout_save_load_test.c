#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include "../include/item.h"
#include "../include/room.h"
#include "../include/player.h"
#include "../include/fileio.h"

int main(int argc, char *argv[]) {
    
    srand(time(0));

    int allocRoomSize = 10; // initial rooms alloc size
    int roomCount = 0;
    Room** rooms = malloc(sizeof(Room*) * allocRoomSize);
    if (rooms == NULL){
        perror("Could not allocate memory for rooms");
        return EXIT_FAILURE;
    }
    
    char layoutStateFPath[] = {"tests/saved_test_games/layout_save_load_test_output.txt"};
    char expectedLayoutStateFPath[] = {"tests/saved_test_games/layout_save_load_test_expected.txt"};

    
    // Create items to be added to the room
    Item* item1 = Item_construct(0);
    Item* item2 = Item_construct(1);
    Item* item3 = Item_construct(2);

    ItemNode* itemsHead = NULL;
    ItemList_insert(&itemsHead, item1);
    ItemList_insert(&itemsHead, item2);
    ItemList_insert(&itemsHead, item3);

    // Create events to be added to the room
    EventNode* events = NULL;
    EventList_insert(&events, FIRE_TRAP);
    EventList_insert(&events, PUZZLE);
    EventList_insert(&events, BLOCKED);
    EventList_insert(&events, LOCKED);


    // Create a room and add it to the rooms array
    Room* room = Room_construct(0, NULL, NULL, NULL, NULL, events, itemsHead);
    if (room != NULL){
        rooms[room->ID] = room;
        roomCount++;
    }
    displayRoom(room);
    // Create another room and add it to the rooms array
    Room* room1 = Room_construct(1, NULL, NULL, NULL, NULL, NULL, NULL);
    if (room1 != NULL){
        room1->north = room;
        room->south = room1;
        rooms[room1->ID] = room1;
        roomCount++;
    }
    displayRoom(room1);


    // save the layout to a file
    int saveStatus = saveLayout(layoutStateFPath, rooms, roomCount);
    if (saveStatus != EXIT_SUCCESS) {
        printf("Error saving layout to file.\n");
        return EXIT_FAILURE;
    }




    // Load the layout from the file
    Room** loadedRooms = NULL;
    int loadedRoomCount = 0;
    int loadedAllocRoomSize = 0;
    bool loadStatus = loadLayout(layoutStateFPath, &loadedRooms, &loadedRoomCount, &loadedAllocRoomSize);
    if (loadStatus != true) {
        printf("Error loading layout from file.\n");
        return EXIT_FAILURE;
    }

    // compare the loaded layout to what was saved
    if (loadedRoomCount != roomCount) {
        printf("Loaded room count does not match saved room count.\n");
        return EXIT_FAILURE;
    }

    printf("\nLoaded room count: %d\n", loadedRoomCount);
    displayRoom(loadedRooms[0]);
    displayRoom(loadedRooms[1]);

    // compare room 0
    if (loadedRooms[0]->ID != rooms[0]->ID ||
        loadedRooms[0]->north != NULL ||
        loadedRooms[0]->south == NULL ||
        loadedRooms[0]->west != NULL ||
        loadedRooms[0]->east != NULL) {
        printf("Loaded room 0 does not match saved room 0.\n");
        return EXIT_FAILURE;
    }
    // compare room 1
    if (loadedRooms[1]->ID != rooms[1]->ID ||
        loadedRooms[1]->north == NULL ||
        loadedRooms[1]->south != NULL ||
        loadedRooms[1]->west != NULL ||
        loadedRooms[1]->east != NULL) {
        printf("Loaded room 1 does not match saved room 1.\n");
        return EXIT_FAILURE;
    }
    // compare items in room 0
    ItemNode* currItem = loadedRooms[0]->items;
    ItemNode* currItem2 = rooms[0]->items;
    for (int i = 0; i < itemListCount(loadedRooms[0]->items); i++) {
        if (currItem == NULL) {
            printf("Loaded item %d in room 0 is NULL.\n", i);
            return EXIT_FAILURE;
        }
        if (currItem->data->ID != currItem2->data->ID) {
            printf("Loaded item %d in room 0 does not match saved item %d.\n", i, currItem2->data->ID);
            return EXIT_FAILURE;
        }
        currItem = currItem->next;
        currItem2 = currItem2->next;
    }
    // compare items in room 1
    currItem = loadedRooms[1]->items;
    currItem2 = rooms[1]->items;
    for (int i = 0; i < itemListCount(loadedRooms[1]->items); i++) {
        if (currItem == NULL) {
            printf("Loaded item %d in room 1 is NULL.\n", i);
            return EXIT_FAILURE;
        }
        if (currItem->data->ID != currItem2->data->ID) {
            printf("Loaded item %d in room 1 does not match saved item %d.\n", i, currItem2->data->ID);
            return EXIT_FAILURE;
        }
        currItem = currItem->next;
        currItem2 = currItem2->next;
    }
    // compare events in room 0
    EventNode* currEvent = loadedRooms[0]->events;
    EventNode* currEvent2 = rooms[0]->events;
    for (int i = 0; i < eventListCount(loadedRooms[0]->events); i++) {
        if (currEvent == NULL) {
            printf("Loaded event %d in room 0 is NULL.\n", i);
            return EXIT_FAILURE;
        }
        if (currEvent->data != currEvent2->data) {
            printf("Loaded event %d in room 0 does not match saved event %d.\n", i, currEvent2->data);
            return EXIT_FAILURE;
        }
        currEvent = currEvent->next;
        currEvent2 = currEvent2->next;
    }
    // compare events in room 1
    currEvent = loadedRooms[1]->events;
    currEvent2 = rooms[1]->events;
    for (int i = 0; i < eventListCount(loadedRooms[1]->events); i++) {
        if (currEvent == NULL) {
            printf("Loaded event %d in room 1 is NULL.\n", i);
            return EXIT_FAILURE;
        }
        if (currEvent->data != currEvent2->data) {
            printf("Loaded event %d in room 1 does not match saved event %d.\n", i, currEvent2->data);
            return EXIT_FAILURE;
        }
        currEvent = currEvent->next;
        currEvent2 = currEvent2->next;
    }





    // free memory
    freeRooms(&rooms, &roomCount);
    freeRooms(&loadedRooms, &loadedRoomCount);


    // compare the saved layout file to the expected layout file
    FILE* savedFile = fopen(layoutStateFPath, "r");
    FILE* expectedFile = fopen(expectedLayoutStateFPath, "r");
    if (savedFile == NULL || expectedFile == NULL) {
        printf("Error opening saved or expected layout file.\n");
        return EXIT_FAILURE;
    }
    char savedLine[256];
    char expectedLine[256];
    while (fgets(savedLine, sizeof(savedLine), savedFile) != NULL &&
           fgets(expectedLine, sizeof(expectedLine), expectedFile) != NULL) {
        if (strcmp(savedLine, expectedLine) != 0) {
            printf("Saved layout file does not match expected layout file.\n");
            fclose(savedFile);
            fclose(expectedFile);
            return EXIT_FAILURE;
        }
    }
    fclose(savedFile);
    fclose(expectedFile);
    printf("Saved layout file matches expected layout file.\n");

    printf("Test passed!\n");
    return EXIT_SUCCESS;    
}