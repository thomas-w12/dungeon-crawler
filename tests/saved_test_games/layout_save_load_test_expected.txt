2
0,Trap,This is a trap room,-1,1,-1,-1,3,5,1,0,1,2,
1,Normal,This is a normal room,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,
